### ISSUE TITLE
[BUG] TypeMismatch: Expected createdAt to be of type date, instead found type object

### ISSUE THREAD

**Issue by AllanOricil** [dynamoose/dynamoose#1037]
### Summary:

CloudWatch log
![image](https://user-images.githubusercontent.com/55927613/99904318-0be88f00-2cc2-11eb-8f32-fa84266c248f.png)

Doc on Dynamodb
![image](https://user-images.githubusercontent.com/55927613/99904299-f5dace80-2cc1-11eb-8447-b83a0fe5d413.png)

#### Schema
```js
const dynamoose = require("dynamoose");
const schema = new dynamoose.Schema({
    email: {
        type: String,
        hashKey: true
    },
    github: Object,
    name: String
}, {
    saveUnknown: true,
    timestamps: true
});

module.exports = dynamoose.model('User', schema);
```

#### General
```js
const { User } = require('../db');
const user = await User.query('email').eq(email).exec();
```

### Current output and behavior (including stack trace):
Cant query from dynamodb

### Expected output and behavior:
I should be able to query from Dynamodb when I pass the pk value

### Environment:

Operating System: NA - Lambda Function
Operating System Version: NA
Node.js version (`node -v`): v10.19.0
NPM version: (`npm -v`): 6.14.8
Dynamoose version: 2.3.0

### Other:
- [x ] I have read through the Dynamoose documentation before posting this issue
- [x ] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [ ] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [x ] I have tested the code provided and am confident it doesn't work as intended
- [ ] I have filled out all fields above
- [ ] I am running the latest version of Dynamoose

**Comment by AllanOricil** [comment]
I changed to the aws sdk and it worked without a problem the query.  I will work with the aws sdk for now and when this is fixed I might change to your lib again. Thanks

````js
        var queryParams = {
            TableName : process.env.USERS_TABLE,
            Key: { email },
        };
        const data = await docClient.get(queryParams).promise();
        console.log(data.Item);
````

**Comment by fishcharlie** [comment]
@AllanOricil Right now I don't have enough information to be able to reproduce this and debug it.

Just for context. I ran the following code.

```js
(async () => {
	const dynamoose = require("dynamoose");
	dynamoose.aws.ddb.local();
	const schema = new dynamoose.Schema({
		email: {
			type: String,
			hashKey: true
		},
		github: Object,
		name: String
	}, {
		saveUnknown: true,
		timestamps: true
	});

	const User = dynamoose.model('User', schema);

	await User.create({
		"email": "test@test.com",
		"github": {
			"credentials": {
				"unknown": "unknown"
			},
			"user": {
				"unknown": "unknown"
			}
		}
	});

	console.log("Created user");

	// Not sure what the variable `email` is set to in your original code so I used `"test@test.com"`
	const user = await User.query('email').eq("test@test.com").exec();
	console.log("Query complete");
	console.log(user);
})();
```

And got the following result:

```
charliefish@Charlies-MacBook-Pro tmp2 % node index.js
Created user
Query complete
[
  Document {
    createdAt: 2020-11-22T16:23:13.519Z,
    github: { credentials: [Object], user: [Object] },
    email: 'test@test.com',
    updatedAt: 2020-11-22T16:23:13.519Z
  },
  lastKey: undefined,
  count: 1,
  queriedCount: undefined,
  timesQueried: 1,
  populate: [Function: PopulateDocuments],
  toJSON: [Function: documentToJSON]
]
```

From my perspective, everything is working fine. Please submit an MCVE as described in the contribution guidelines so I can debug this further.

**Comment by AllanOricil** [comment]
> @AllanOricil Right now I don't have enough information to be able to reproduce this and debug it.
> 
> Just for context. I ran the following code.
> 
> ```js
> (async () => {
> 	const dynamoose = require("dynamoose");
> 	dynamoose.aws.ddb.local();
> 	const schema = new dynamoose.Schema({
> 		email: {
> 			type: String,
> 			hashKey: true
> 		},
> 		github: Object,
> 		name: String
> 	}, {
> 		saveUnknown: true,
> 		timestamps: true
> 	});
> 
> 	const User = dynamoose.model('User', schema);
> 
> 	await User.create({
> 		"email": "test@test.com",
> 		"github": {
> 			"credentials": {
> 				"unknown": "unknown"
> 			},
> 			"user": {
> 				"unknown": "unknown"
> 			}
> 		}
> 	});
> 
> 	console.log("Created user");
> 
> 	// Not sure what the variable `email` is set to in your original code so I used `"test@test.com"`
> 	const user = await User.query('email').eq("test@test.com").exec();
> 	console.log("Query complete");
> 	console.log(user);
> })();
> ```
> 
> And got the following result:
> 
> ```
> charliefish@Charlies-MacBook-Pro tmp2 % node index.js
> Created user
> Query complete
> [
>   Document {
>     createdAt: 2020-11-22T16:23:13.519Z,
>     github: { credentials: [Object], user: [Object] },
>     email: 'test@test.com',
>     updatedAt: 2020-11-22T16:23:13.519Z
>   },
>   lastKey: undefined,
>   count: 1,
>   queriedCount: undefined,
>   timesQueried: 1,
>   populate: [Function: PopulateDocuments],
>   toJSON: [Function: documentToJSON]
> ]
> ```
> 
> From my perspective, everything is working fine. Please submit an MCVE as described in the contribution guidelines so I can debug this further.

With a local DynamoDb (using localstack) the code works too. The problem is when we use DynamoDB instance on Amazon. For some reason the "createdAt", which is saved as a Number, is throwing an exception when it is deserialized. But this only happens when I use aws.

**Comment by AllanOricil** [comment]
@fishcharlie you can use this SAM App to speed things up https://github.com/AllanOricil/github-auth-serverless
Just change the parts where Im now using the native aws sdk to use Dynamoose, and you will see the error.

**Comment by fishcharlie** [comment]
@AllanOricil

I will note. It is expected to save as a number. DynamoDB has no concept of a date. So behind the scenes for the `Date` type in Dynamoose, Dynamoose will serialize and deserialized it automatically for you.

But you said my code works in DynamoDB Local, but not the AWS DynamoDB servers. I commented out `dynamoose.aws.ddb.local();`. And got the same result, where it works. So I'm still not seeing any error message.

You can see the entry in [Dynobase](https://dynobase.dev/?ref=dynamoose) below.

<img width="1135" alt="image" src="https://user-images.githubusercontent.com/860375/99910139-a5577700-2ca9-11eb-8c36-9e3275596a26.png">

---

As I mentioned:

> Please submit an MCVE as described in the contribution guidelines so I can debug this further.

**Comment by AllanOricil** [comment]
I gave you everything I had to show the bug. What else can I give you to determine the cause? The error comes from line 302 in the Document class, so there is nothing I can do. The only thing I'm sure is that I just changed the lines for the query to use AWS SDK and the bug was gone, so my code has nothing wrong.

**Comment by fishcharlie** [comment]
@AllanOricil 

> What else can I give you to determine the cause?

As I've said now multiple times, please submit an MCVE (Minimal Complete Verifiable Example) so I can debug further. You can find information about that in the contributing guidelines. But I'll spell it out here as well.

Basically **create a new project**, and figure out what the **least** amount of code is to reproduce the problem. It should be complete, meaning it should include everything including your `node_modules` folder. The only thing you can leave out is the region, and IAM keys. And it should be verifiable. I should be able to do the following:

1. Download and uncompress the ZIP file you upload to a comment on this issue
2. Open Terminal
3. Export my AWS region and IAM keys as environment variables
4. Run `node index.js` within the folder you provide in the ZIP file
5. See the error you are getting

That is how easy it should be. You gave me a link to a repo that has a bunch of SAM stuff, with unit tests, and you also mentioned `Just change the parts where Im now using the native aws sdk to use Dynamoose`. That isn't helpful **at all**. For all I know the way you went from AWS SDK to Dynamoose messed something up. Maybe you didn't do it right? I have no idea. I don't think that is what's going on here. But obviously there is a bigger problem since my code works and yours doesn't. I also don't wanna have to dig through a bunch of code that I don't understand. That takes a lot of my time.

Another option, would be to provide me a complete example like I did. I pasted all my code I ran in the comment. And as you can see, it did everything. It created the item, and got to the state that you discussed in your original bug report. Then it tried to query to reproduce the problem (without success).

The problem with your original issue is that it's not complete or verifiable. I don't have insight into what `credentials` is, I don't have insight into what `user` is. I don't have insight into what the `email` variable is. I don't have insight into what other items/documents you might have in your table. It might be minimal, but I'm not sure. Does this problem happen if you try to reproduce it without the `credentials` or `user` stuff in the documents & schema? Is it those that are causing problems? If so, it's minimal. If the problem exists without that extra stuff, it's not minimal. Like for all I know there is something in `credentials` or `user` that is causing the data to get corrupted or something. I have no idea. But if you create an MCVE and try things, it helps narrow the problem a lot more. It might even lead to you figuring out the problem on your own.

Then the problem with your repo link you gave me is it's not minimal or complete. It might be verifiable, but I'm not sure. It has a bunch of code that is irrelevant to me and doesn't relate to the problem we are facing, along with a bunch of extra steps to reproduce it. It's also not complete since you said `Just change the parts where Im now using the native aws sdk to use Dynamoose`. That's not complete. I'm not going to spend the time doing that work for you.

---

I **fully** understand that creating a MCVE is extremely time consuming. But I've found, when doing it myself, it helps me solve the issue myself the majority of the time. And if not, it makes it **extremely** easy for others to help debug the problem. It will save us from having to go back and forth a bunch of times like we are now.

---

> The error comes from line 302 in the Document class, as indicated by cloud watch logs

I understand.

Once you provide me with that MCVE, I will run the steps I listed above. And **hopefully** I'll get the same error you are getting. Then, I will go through, and set it up to use my local Dynamoose version. I understand the Dynamoose code very well. I will go through and set breakpoints in the Dynamoose code, and figure out exactly why it's failing. Then I'll fix the issue, or report back with a solution.

But at this point I can't even get the same error you are getting. So I can't test things or use breakpoints in the Dynamoose code to try to debug.

I **really want to help you**, but you have to help me help you. I can't help at all until I can reproduce this problem on my local machine.

---

> The only thing I'm sure is that I just changed the lines for the query to use AWS SDK and the bug was gone, so my code has nothing wrong

That isn't necessarily true. There could be a problem with how your code uses or interacts with Dynamoose. But, you are correct, it could be a problem with Dynamoose as well. Which is why I need to reproduce it and debug it to fix any problems.

**Comment by fishcharlie** [comment]
@AllanOricil One thing I just thought of. In **all** of your items/documents in that table. Do you have any `NULL` DynamoDB types? Dynamoose doesn't support the DynamoDB `NULL` type currently (#951). And the error you are having has been shown to be a result of that in the past. I'm aware the error message is pretty bad. But the amount of effort it'd require to make the error message better is the same amount of effort as adding support for `NULL` types.

**Comment by AllanOricil** [comment]
> @AllanOricil One thing I just thought of. In **all** of your items/documents in that table. Do you have any `NULL` DynamoDB types? Dynamoose doesn't support the DynamoDB `NULL` type currently (#951). And the error you are having has been shown to be a result of that in the past. I'm aware the error message is pretty bad. But the amount of effort it'd require to make the error message better is the same amount of effort as adding support for `NULL` types.

@fishcharlie I only had one single document in the table. The only one with the email "allanoricilcos@outlook.com". All the 4 attributes on that doc were filled with something too.

**Comment by fishcharlie** [comment]
@AllanOricil How about the properties within user and credentials? Your screenshot didn't show any of that information.

If there is still no null property types in those nested objects please submit that MCVE.

**Comment by AllanOricil** [comment]
> @AllanOricil How about the properties within user and credentials? Your screenshot didn't show any of that information.
> 
> If there is still no null property types in those nested objects please submit that MCVE.

This is what Im saving:

````json
{
  "email": "allanoricilcos@outlook.com",
  "github": {
    "credentials": {
      "access_token": "REMOVED_FOR_SAFETY",
      "expires_in": "28800",
      "refresh_token": "REMOVED_FOR_SAFETY",
      "refresh_token_expires_in": "15638400",
      "scope": "",
      "token_type": "bearer"
    },
    "user": {
      "avatar_url": "https://avatars0.githubusercontent.com/u/55927613?v=4",
      "bio": null,
      "blog": "",
      "company": null,
      "created_at": "2019-09-28T18:49:04Z",
      "email": null,
      "events_url": "https://api.github.com/users/AllanOricil/events{/privacy}",
      "followers": 4,
      "followers_url": "https://api.github.com/users/AllanOricil/followers",
      "following": 5,
      "following_url": "https://api.github.com/users/AllanOricil/following{/other_user}",
      "gists_url": "https://api.github.com/users/AllanOricil/gists{/gist_id}",
      "gravatar_id": "",
      "hireable": true,
      "html_url": "https://github.com/AllanOricil",
      "id": 55927613,
      "location": "Brazil",
      "login": "AllanOricil",
      "name": "Allan Oricil",
      "node_id": "MDQ6VXNlcjU1OTI3NjEz",
      "organizations_url": "https://api.github.com/users/AllanOricil/orgs",
      "public_gists": 0,
      "public_repos": 15,
      "received_events_url": "https://api.github.com/users/AllanOricil/received_events",
      "repos_url": "https://api.github.com/users/AllanOricil/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/AllanOricil/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/AllanOricil/subscriptions",
      "twitter_username": null,
      "type": "User",
      "updated_at": "2020-11-22T20:04:48Z",
      "url": "https://api.github.com/users/AllanOricil"
    }
  },
  "name": "Allan Oricil"
}

````

**Comment by fishcharlie** [comment]
@AllanOricil There you go. `user.bio`, `user.company`, `user.email`, list goes on. All `null`. Next time, when submitting issues, please be sure to read through the contributing guidelines. It's very clear about MCVE. In the future I will be closing issues that don't have enough information for me to debug.

And please, when I ask "Do you have any NULL DynamoDB types?", the answer is "Yes". No need to waste time with back and forth answers. Please next time, just provide the information I need so I can focus on helping as many users as possible and not wasting time.

I'm all for helping as much as I can. But help me help you by being efficient.

---

As for this issue, I'm 99% sure it's a duplicate of #951. In order to fix it, you could remove those `null` properties from your document/item and make sure to delete them before saving any document/item as well. One way to do that with your setup is to set them to `undefined` before saving.

That should fix this issue. Now, in your use case, I'm not sure what the difference would be between deleting them and keeping them as `null`. Even if #951 was implemented, I think you might still want to delete them to save on document/item size. That would save money at scale. Why store `user.bio: null` when you could just not store anything? In your use case, doesn't `null` equal nothing? So not saving it at all would help reduce document/item size. Downside is that it would no longer be the same object you get back from GitHub (I think that is the source of the data???). Not sure why that'd be a problem tho.

If you want to store those null values, you'll have to wait until #951 gets implemented. I'll do my best to prioritize it and get it done soon. No ETA on when that'll happen tho.

---

Hopefully this helps resolve the issue! If it doesn't, please submit that MCVE I mentioned above.

**Comment by AllanOricil** [comment]
@fishcharlie I changed to aws sdk. The reason for the issue is probably 99% related to that #951, as you said, but the error message is different. Better to leave this open so other people can search this error message and not waste your time. This will help you to be much more efficient ;)
They will also be able to understand that to fix it is just a matter of learning how to use the high level sdk calls.
Once you fix that one then you can come and close this, as the error message will never appear again.
Anyway, thanks for the help.

**Comment by fishcharlie** [comment]
@AllanOricil This issue will be closed out automatically as stale after it's inactive for a certain amount of time. Users can still view this issue after it's closed and see it's a duplicate of #951.

**Comment by yosle** [comment]
I know this is a bit late. But   there was no clear solution for the problem yet. So here I go:  I had the same problem, same error message. I was using Local instance of DynamoDB so I deleted the tables using NoSQLWorkbench. Seems that I have made a change in the createdAt data type  and it was conflicting with the previously saved schema.
